### ISSUE TITLE
Feature request: Function equivalent to elixir's `Task.async_stream`

### ISSUE THREAD

**Issue by TanklesXL** [gleam-lang/otp#32]
As previously discussed, this would be a nice  & useful function to have.

**Comment by lpil** [comment]
Great suggestion. Initial thoughts:

The name isn't very clear- what would be clearer?

The Elixir function is lazy, which makes me think it spawns a process per element requested in order to avoid leaking processes if you never reach the end. Would a version that always processes the entire collection but uses a work pool be advantageous?

**Comment by TanklesXL** [comment]
In keeping with the original intent of `async_stream` maybe a name like `concurrent_map`, `parallel_map`? Might be nice to have an alternative that works on a list of thunks as well maybe?

**Comment by TanklesXL** [comment]
Dropping this link to the [parmap](https://github.com/rdicosmo/parmap) lib for reference.

Seems like a good idea to let the caller select the number of workers, allows for either a smaller pool of processes to each handle multiple items or for 1-1 items to processes.

**Comment by lpil** [comment]
`parallel_map` sounds nice and descriptive. What do you mean about thunks?

**Comment by TanklesXL** [comment]
With regards to thunks I was mostly looking for a way to run a list of `fn() -> a` concurrently without needing to do something like 
```
l
|> list.map(task.async)
|> list.map(task.await...)
```

**Comment by TanklesXL** [comment]
Maybe an api like ocaml's `parmap` lib might be nice? 

https://rdicosmo.github.io/parmap/doc/parmap/Parmap/index.html

**Comment by lpil** [comment]
I'm not sure how that would translate into Gleam as we don't have module functors. Keen to see what you have in mind

**Comment by TanklesXL** [comment]
Oh that's a good point, hmm this definitely requires more thought. 

Might be worth opening a discussion around this issue? I'd be interested in seeing if anyone has ideas

**Comment by jrstrunk** [comment]
Is an implementation like this something you were looking for? I could add documentation, tests, clean it up a little, and make any suggested changes if so. This will process the list of functions in 5 tasks concurrently, each pulling from the list of functions to execute until there are none. The entire pool has to be complete by 60 seconds in this example.

```gleam
let secret = "1234"
let inputs = list.range(0, 1000)

inputs
|> list.map(fn(param) {
  fn() {
    process_function(param, secret)
  }
})
|> task.async_pool(5)
|> task.await(60_000)
```
An alternative interface that would be less flexible, but maybe a little nicer and closer to the mentioned elixir function would be something like:
```gleam
let secret = "1234"
let inputs = list.range(0, 1000)

inputs
|> task.async_pool(process_function(_, secret), 5)
|> task.await(60_000)
```

Here is the implementation of the first example that I wrote for one of my own programs:
```gleam
import gleam/bool
import gleam/erlang/process
import gleam/list
import gleam/otp/actor
import gleam/otp/task

pub fn async_pool(
  make_tasks: List(fn() -> a),
  num_workers: Int,
) -> task.Task(List(a)) {
  use <- bool.guard(when: num_workers <= 0, return: task.async(fn() { [] }))

  let assert Ok(sync_queue) = actor.start(make_tasks, handle_sync_queue)

  use <- task.async
  list.range(1, num_workers)
  |> list.map(fn(_worker_id) {
    task.async(fn() {
      let res = process_sync_queue(sync_queue, [])
      res
    })
  })
  |> list.map(task.await_forever)
  |> list.flatten
  // Do we need to stop the sync queue actor here?
}

type SyncQueueResponse(a) {
  SyncQueueItem(a)
  SyncQueueEmpty
}

fn handle_sync_queue(reply, queue) {
  case queue {
    [item, ..rest] -> {
      actor.send(reply, SyncQueueItem(item))
      rest
    }
    [] -> {
      actor.send(reply, SyncQueueEmpty)
      []
    }
  }
  |> actor.continue
}

fn process_sync_queue(queue, results) {
  case process.call_forever(queue, fn(a) { a }) {
    SyncQueueItem(item) -> process_sync_queue(queue, [item(), ..results])
    SyncQueueEmpty -> results
  }
}

```

**Comment by lpil** [comment]
That implementation would have fewer than 5 tasks running the majority of the time, we would want to spawn the next task immediately after the previous one finished rather than waiting for the entire batch to finish.

**Comment by jrstrunk** [comment]
If that were true then wouldn't this output
```
Starting  Task 1: 11:19:16.592
Starting  Task 2: 11:19:16.592
Done with Task 2: 11:19:16.700
Starting  Task 3: 11:19:16.700
Done with Task 3: 11:19:16.801
Starting  Task 4: 11:19:16.801
Done with Task 4: 11:19:16.902
Done with Task 1: 11:19:26.599
```
for this program be impossible?
```gleam
pub fn main() {
  [
    fn() {
      io.println(
        "Starting  Task 1: " <> time.now_local() |> time.to_string,
      )
      process.sleep(10_000)
      io.println(
        "Done with Task 1: " <> time.now_local() |> time.to_string,
      )
    },
    fn() {
      io.println(
        "Starting  Task 2: " <> time.now_local() |> time.to_string,
      )
      process.sleep(100)
      io.println(
        "Done with Task 2: " <> time.now_local() |> time.to_string,
      )
    },
    fn() {
      io.println(
        "Starting  Task 3: " <> time.now_local() |> time.to_string,
      )
      process.sleep(100)
      io.println(
        "Done with Task 3: " <> time.now_local() |> time.to_string,
      )
    },
    fn() {
      io.println(
        "Starting  Task 4: " <> time.now_local() |> time.to_string,
      )
      process.sleep(100)
      io.println(
        "Done with Task 4: " <> time.now_local() |> time.to_string,
      )
    },
  ]
  |> taskx.async_pool(2)
  |> task.await_forever
}
```

The first worker in the pool processes task 1 the entire time, while the second worker processes task 2, then 3, then 4, then runs out of tasks to process. The parent task is blocked by all tasks in the pool completing, but each task recursively pulls in new items to process as soon as available.

**Comment by lpil** [comment]
Ah! Sorry! I misread. I saw a list of tasks being mapped and thought it was batching.

This is instead misusing tasks and has very different properties to how tasks are supposed to work, and code that is written to use tasks could fail if run this way.

A task is a _short lived_ process that does 1 brief item of work and then terminates. It is never appropriate to reuse a task process. If you wish for a long lived process that does something other than perform a single CPU bound item of work then you should use actors.

**Comment by jrstrunk** [comment]
Ah I see, good thing to learn! Would an implementation that started an actor which spun up a new task to process every item in the list of inputs, while making sure that there are always exactly `n` tasks running concurrently be a solution?

**Comment by lpil** [comment]
I expect we'd need an implementation of supervised tasks, which I expect would be a totally new implementation under the hood, along with dynamic-supervisor.

Stepping back a bit, I'm not sure why we want this to be specifically related to tasks. Would something based on producer-consumers like Elixir's broadway not be always preferred? Like the Flow and Broadway libraries.

**Comment by jrstrunk** [comment]
I have started to come to the same conclusion: I think we are stepping outside of the fit of the `task` module. I wrote an implementation using actors with no supervisor (because if something fails I just wanted the whole thing to stop), and I gave it the semantics of a "work pool" and "workers" instead of "tasks". The implementation is [here](https://gist.github.com/jrstrunk/c09fe4acf848a2a68eef8ad761eddac6) if anyone is interested. 

I think this functionality is really nice, but needs more consideration where it fits. I am happy with my implementation for my use case, but I don't think I have enough expertise to aid in getting this added as part of the std Gleam package. If anyone else wants to pick up and continue working this problem then I'm sure the ecosystem would appreciate it!
